docs/terrain.md

```diff
@@ -0,0 +1,123 @@
+| Chunks     | [Chunk](#chunk) | As many chunks as needed to fill the blob. Unknown/no maximum amount. |
+
+## Data Types
+
+Terrain data is represented using a variety of different data types. Data described within them follows any conventions set in the [document conventions](#document-conventions).
+
+### Chunk
+
+The `Chunk` type is stored with a dynamic size dependent on the size of the voxels contained within, and the end of its data is marked by reaching its maximum voxel count of 32<sup>3</sup>. Voxels are ascendingly ordered by Y, then Z, then X based on their position in the chunk. Voxels are stored in rows of 32 units on each axis, each representing 4 units in world space.
+
+Voxel data is preceded by the offset between this chunk and the last in the blob, or from `0, 0, 0` if this is the first chunk in the blob. This offset is stored using 3 vectors of `(x: u8, y: u8, z: u8)` using the signedness determination described in [document conventions](#document-conventions), with 0xFF values in _all_ unused offsets indicating a negative sign. All coordinates are in chunk space (increments of 1 chunk), not world space.
+
+For example, given two chunks, one at a position of `(2, 0, 0)` and another at a position of `(4, 0, -1)` in chunk space, the latter's offset would be stored as follows:
+
+1. (0x00, 0x00, 0xFF)
+2. (0x00, 0x00, 0xFF)
+3. (0x00, 0x00, 0xFF)
+4. (0x02, 0x00, 0x01)
+
+| Field Name     | Format          | Value                                                                               |
+| :------------- | :-------------- | :---------------------------------------------------------------------------------- |
+| Signedness     | `[u8; 3]`       | Always the signedness of the following vectors. 0xFF in negative axes.              |
+| Offset (65536) | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 65536 chunks. |
+| Offset (256)   | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 256 chunks.   |
+| Offset (1)     | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being in singular chunks.         |
+| Voxels         | [Voxel](#voxel) | As many voxels as needed to fulfill the maximum amount per chunk.                   |
+
+### Voxel
+
+The `Voxel` type is stored with a dynamic size (between 1-4 bytes) dependent on its set [bitflags](#voxel.flag). Only stored within [chunks](#chunk). A reference for the data contained within voxels [can be found here][Terrain.WriteVoxelChannels]. Empty voxels are stored with their material set to Air.
+
+Voxels are stored within a chunk using [run-length encoding]. A count of `1` would indicate the voxel repeating itself once, meaning 2 of the same voxel in a row would be read when decoded. Occupancy values are stored using the float format described in the [document conventions](#document-conventions).
```

Comment:
When you say "run length encoding" do you mean that the size of the `Count` field increases dynamically or is it just a number between 0 and 255? If it's the latter, this is a misuse of the term and it's confusing.

If it's the former, this needs to be rephrased because it suggests to me that the value is just a simple "this voxel repeats between 0 and 255 times".

Reply:
```
When you say "run length encoding" do you mean that the size of the Count field increases dynamically or is it just a number between 0 and 255? If it's the latter, this is a misuse of the term and it's confusing.
```
This is like the point of run length encoding though. The count is a number between 0 and 255 determining how long a run length is, AKA, on decode, the voxel repeats itself `count` times and should be read as such, while only taking 2-3 bytes of actual storage space.
```
Run-length encoding (RLE) is a form of [lossless data compression](https://en.wikipedia.org/wiki/Lossless_data_compression) in which runs of data (consecutive occurrences of the same data value) are stored as a single occurrence of that data value and a count of its consecutive occurrences, rather than as the original run.
```